app/helpers/pagination_helper.rb

```diff
@@ -13,4 +13,44 @@ def pagination(collection=nil, &block)
     html.html_safe
   end
 
+  # Kaminari generates a URL from the params hash. The URL is, however, not
+  # compatible to loc/ids.
+  # This method builds the loc/id compatible link to the page.
+  def build_link_from_request(kaminari_url)
+    page, per_page = params_from_kaminari_url(kaminari_url)
+
+    query_string_parts = request.env['QUERY_STRING'].
+      split(/;|&/).
+      map { |p| p.split('=') }
+
+    query_string_parts.reject! { |p| p.first == 'page' } if page
+    query_string_parts.reject! { |p| p.first == 'per_page' } if per_page
+
+    query_string_parts << ['page', page] if page
+    query_string_parts << ['per_page', per_page] if per_page && per_page != :default
+
+    query_string_parts.map! { |p| p.join('=') }
+    query_string = query_string_parts.join('&')
+
+    [request.env['REQUEST_PATH'], query_string].compact.join('?')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.